ee/tabby-webserver/src/repositories.rs

```diff
@@ -0,0 +1,125 @@
+use std::{collections::HashMap, path::PathBuf, str::FromStr};
+
+use anyhow::Result;
+use axum::{
+    body::boxed,
+    http::{header, Request, Uri},
+    response::{IntoResponse, Response},
+    Json,
+};
+use filenamify::filenamify;
+use hyper::Body;
+use mime_guess::mime;
+use serde::{Deserialize, Serialize};
+use tabby_common::SourceFile;
+use tokio::sync::OnceCell;
+use tower::ServiceExt;
+use tower_http::services::ServeDir;
+use tracing::error;
+
+pub(crate) static DATASET: OnceCell<HashMap<DatasetKey, SourceFile>> = OnceCell::const_new();
+
+#[derive(Hash, PartialEq, Eq, Debug)]
+pub struct DatasetKey {
+    local_name: String,
+    rel_path: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Repository {
+    name: String,
+    path: Option<String>,
+}
+
+impl Repository {
+    pub fn dataset_key(&self) -> DatasetKey {
+        DatasetKey {
+            local_name: self.name.clone(),
+            rel_path: self.path_str().to_string(),
+        }
+    }
+
+    pub fn name_str(&self) -> &str {
+        self.name.as_str()
+    }
+
+    pub fn path_str(&self) -> &str {
+        self.path.as_deref().unwrap_or("")
+    }
+}
+
+#[derive(Serialize)]
+struct ListDir {
+    entries: Vec<String>,
+}
+
+/// Load dataset
+pub async fn load_dataset() -> Result<()> {
+    // `SourceFile::all()` depends on `std::io`, so it's blocking.
```

Comment:
Maybe refactor SourceFile::all into non-blocking (await style)?

Reply:
(can be individual PR)

Reply:
As I checked, there's no async counterpart of `glob::glob()`.
Even though we open file and create reader asynchronously, there still exists blocking call.